### PR TITLE
redefine basic Equals and hashcode for object rectangle

### PR DESCRIPTION
**Pull request by fyarkiy**

### CODE REVIEW

File: src/main/java/core/basesyntax/Rectangle.java [hunk 1]
```diff
@@ -27,4 +27,39 @@ public Integer getLength() {
...
     }
+
+    public boolean equals(Object object) {
+        if (this == object) {
+            return true;
+        }
+        if (object == null) {
+            return false;
+        }
+        Rectangle rectangle = (Rectangle) object;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Rectangle.java]: will cast be successful if `object` is not a Rectangle?

File: src/main/java/core/basesyntax/Rectangle.java [hunk 2]
```diff
@@ -27,4 +27,39 @@ public Integer getLength() {
...
+                    && ((getColor() == null && rectangle.getColor() == null)
+                    || (getColor().equals(rectangle.getColor()))));
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 23;
+        if (getWidth() != null) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Rectangle.java]: do we need to use getters from the same class?

File: src/main/java/core/basesyntax/Rectangle.java [hunk 3]
```diff
@@ -27,4 +27,39 @@ public Integer getLength() {
...
+        }
+        Rectangle rectangle = (Rectangle) object;
+
+        if (object.getClass().equals(Rectangle.class)) {
+            return (((getWidth() == null && rectangle.getWidth() == null)
+                    || getWidth().equals(rectangle.getWidth()))
+                    && ((getLength() == null && rectangle.getWidth() == null)
+                    || getLength().equals(rectangle.getLength()))
+                    && ((getColor() == null && rectangle.getColor() == null)
+                    || (getColor().equals(rectangle.getColor()))));
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Rectangle.java]: what if width == null bit rectangle.width != null?
`getWidth().equals(rectangle.getWidth())` will this code work?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/Rectangle.java (hunk 3 above)
Comment (fyarkiy) [reply src/main/java/core/basesyntax/Rectangle.java]: No. made changes...

File: src/main/java/core/basesyntax/Rectangle.java [hunk 4]
```diff
@@ -27,4 +27,48 @@ public Integer getLength() {
...
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 23;
+        if (width != null) {
+            result = 31 * result * width;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Rectangle.java]: let's use width.hashCode() here

File: src/main/java/core/basesyntax/Rectangle.java [hunk 5]
```diff
@@ -27,4 +27,48 @@ public Integer getLength() {
...
+        }
+        if (length != null) {
+            result = 31 * result * length;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Rectangle.java]: same

File: src/main/java/core/basesyntax/Rectangle.java [hunk 6]
```diff
@@ -27,4 +27,48 @@ public Integer getLength() {
...
+                    || (getColor() != null && rectangle.getColor() == null)) {
+                return false;
+            }
+
+            return (((getWidth() == null && rectangle.getWidth() == null)
+                    || getWidth().equals(rectangle.getWidth()))
+                    && ((getLength() == null && rectangle.getWidth() == null)
+                    || getLength().equals(rectangle.getLength()))
+                    && ((getColor() == null && rectangle.getColor() == null)
+                    || (getColor().equals(rectangle.getColor()))));
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/Rectangle.java]: move all this logic to a single return statement please
see checklist: While comparing fields at equals, use single return statement for all fields(e.g. return check1 && check2 && check3;)

also, don't use getters - use fields instead. you have direct access to the fields from the class where they are declared

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request closed without merging**